### PR TITLE
EVG-3132 Add parsing for new `go test -json` format

### PR DESCRIPTION
**Pull request by richardsamuels**
Patch that shows what test results look like: https://evergreen-staging.corp.mongodb.com/version/5b34fa9797b1d34729506b79

### CODE REVIEW

File: command/results_go_test2json.go [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+		TestFile:  key,
+		LineNum:   test2JSON.StartLine,
+		Status:    evergreen.TestFailedStatus,
+		StartTime: float64(test2JSON.StartTime.Unix()),
+		EndTime:   float64(test2JSON.EndTime.Unix()),
+	}
+	switch test2JSON.Status {
+	case "pass":
+		result.Status = evergreen.TestSucceededStatus
+	case "skip":
```
Comment (tychoish) [inline command/results_go_test2json.go]: didn't you define constants for this?

File: command/results_go_test2json.go [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+	// run    - the test has started running
+	// pause  - the test has been paused
+	// cont   - the test has continued running
+	// pass   - the test passed
+	// bench  - the benchmark printed log output but did not fail
+	// fail   - the test or benchmark failed
+	// output - the test printed output
+	// Additionally, the following is generated by go 1.10.3 but is undocumented:
+	// skip   - the test was skipped
+	Action string
```
Comment (tychoish) [inline command/results_go_test2json.go]: should we make these a different type just to clarify things?

File: command/results_go_test2json.go [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+	"github.com/pkg/errors"
+)
+
+var (
+	testStartRegex = regexp.MustCompile(`^=== RUN\s+.*$`)
+)
+
+const (
+	benchFailPrefix = "--- FAIL: %s"
+	benchSkipPrefix = "--- SKIP: %s"
```
Comment (tychoish) [inline command/results_go_test2json.go]: surely these constants and the regex above aren't relevant to this format?

File: command/results_go_test2json.go [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+
+func goTest2JSONFactory() Command          { return &goTest2JSONCommand{} }
+func (c *goTest2JSONCommand) Name() string { return "gotest.parse_json" }
+
+func (c *goTest2JSONCommand) ParseParams(params map[string]interface{}) error {
+	if err := mapstructure.Decode(params, c); err != nil {
+		return errors.Wrapf(err, "error decoding '%s' params", c.Name())
+	}
+
+	if len(c.Files) == 0 {
```
Comment (tychoish) [inline command/results_go_test2json.go]: do we want to provide/support gitignore (or even regular) file matching/expansion? the commands that work on files are sort of devided on the question.

File: command/results_go_test2json.go (hunk 3 above)
Comment (richardsamuels) [reply command/results_go_test2json.go]: See point 6 above the struct. test2json does not correctly tag benchmark test names

**Comment by richardsamuels** [comment]
@tychoish 
1. By public, do you mean as its own repo or do you mean just as a subpackage under util?
2. Start/end events are not on the same line, or in any defined order when parallelising tests
3. go test -json merges stdout/stderr. Each line of text emits a single test event, so when a panic report is generated, a bunch of test events with action "output" are emitted.

**Pull request merged**